Bug fix, timestamp fix, documentation update

roycamp opened the pull request:
Fixed serverIndex Null Pointer bug
Preserve flume timestamp if it exists
Updated documentation

Commit:
Removing serverIndex as this.servers.get(this.serverIndex++) causes Array Index Out of Bounds when more than one server is supplied.  This increment functionality isn't needed since server failures result in the server being removed from the server list and re-added after the retry period.

Commit:
Added example usage for multiple servers.  
Added Troubleshooting section.

Commit:
Formatting fixes

Commit:
Preserve timestamp from when the event was generated if it exists.

Commit:
Fix key generation to use timestamp of event; still handles empty/missing time.

thobbs commented on src/java/org/apache/cassandra/plugins/CassandraClient.java:
```diff
@@ -165,7 +163,7 @@ String get() throws NoServersAvailableException {
         throw new NoServersAvailableException();
       }
-      return this.servers.get(this.serverIndex++);
+      return this.servers.get(0);
```
So, just incrementing serverIndex without making sure it hasn't gone off the end of the list was definitely broken, but we do want to cycle through the servers in our list.  If we always use 0, that won't happen.

I can make the change, but we probably just need to modulus by serverList.size().

roycamp replied:
My understanding is that ServerSet.get() is only called from CassandraClient.open().  CassandraClient.open() is only called during start up and  when a Cassandra exception is thrown which results in ServerSet.markDead().  markDead() removes the server from the servers array thus causing the remaining servers to shift up in the array.  When the markDead() timeout expires, the server is added onto the end of the servers array.  This is how the cycling of servers is done.  Is there a scenario where open() is called which I am missing?

thobbs replied:
Ah, you're totally correct.  Thanks.

thobbs commented on src/java/org/apache/cassandra/plugins/SimpleCassandraSink.java:
```diff
@@ -63,7 +63,10 @@ public void open() throws IOException {
 
-    long timestamp = System.currentTimeMillis() * MILLI_TO_MICRO;
+    // Preserve timestamp from when the event was generated 
+    long timestamp = event.getTimestamp();  
```
The main problem with doing this is that the TimeUUID we generate just below this won't use the same timestamp, so columns in the index rows won't necessarily match up with the time range in the row key.

It's possible to create a TimeUUID given a timestamp, but we would need to swap out UUID libraries and use something like this: https://github.com/rantav/hector/blob/master/core/src/main/java/me/prettyprint/cassandra/utils/TimeUUIDUtils.java

roycamp replied:
Ah, good catch.  I only accounted for the row key not the column TimeUUID.  I will commit a fix.

Commit:
Merge remote branch 'original/master'

Commit:
Adding new dependency for improved UUID functionality
http://wiki.apache.org/cassandra/FAQ#working_with_timeuuid_in_java
http://johannburkard.de/software/uuid/

Commit:
Added TimeUUIDUtils from Hector with minor modifications. Updated Sink to generate column TimeUUID from incoming timestamp instead of current time.

Commit:
Merge branch 'master' of github.com:roycamp/flume-cassandra-plugin

Commit:
Actually including the TimeUUIDUtils file this time...

roycamp commented:
I merged the most recent updates (maven support) -- hopefully did this right.
Added TimeUUID generation support for column names based on the event timestamp.

Commit:
Convert java.util.uuid to byte array

roycamp commented:
Do I need to include the UUID jar (from http://johannburkard.de/software/uuid/) in the lib folder or will the maven dependency handle it?

thobbs commented:
If the build were set up to make a single uber-jar, then we could forgo adding it to lib (and remove the requirement for the user to add multiple jars to the classpath).  I'll merge this in and then see if I can make that happen.

Thanks!
